File: graphs/word_ladder.rb
# @param {String} begin_word
# @param {String} end_word
# @param {String[]} word_list
# @return {Integer}
def ladder_length(begin_word, end_word, word_list)
    set = Set.new(word_list)
    queue = [[begin_word, 1]]

    while !queue.empty?
        word, length = queue.shift
        return length if word == end_word

        word.chars.each_with_index do |_, i|
            ('a'..'z').each do |ch|
                new_word = word.dup
                new_word[i] = ch
                if set.include?(new_word)
                    queue << [new_word, length + 1]
                    set.delete(new_word)
                end
            end
        end
    end
    0
end

File: graphs/word_ladder.py
from collections import deque

class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        st = set(wordList)
        queue = deque([(beginWord, 1)])
        
        while queue:
            word, length = queue.popleft()
            if word == endWord:
                return length

            for i in range(len(word)):
                temp = list(word)
                for ch in "abcdefghijklmnopqrstuvwxyz":
                    temp[i] = ch
                    new_word = "".join(temp)
                    if new_word in st:
                        queue.append((new_word, length + 1))
                        st.remove(new_word)

        return 0